torch_np/_reductions.py

```diff
@@ -60,136 +59,219 @@ def _atleast_float(dtype, other_dtype):
 
     if dtype == torch.bool:
         dtype = _dtypes_impl.default_dtypes.int_dtype
 
     if axis is None:
-        result = tensor.sum(dtype=dtype)
+        result = a.sum(dtype=dtype)
     else:
-        result = tensor.sum(dtype=dtype, dim=axis)
+        result = a.sum(dtype=dtype, dim=axis)
 
     return result
 
 
-@emulate_keepdims
-@deco_axis_expand
-def prod(tensor, axis=None, dtype=None, initial=None, where=None):
+@_deco_axis_expand
+def prod(
+    a: ArrayLike,
+    axis: AxisLike = None,
+    dtype: Optional[DTypeLike] = None,
+    out: Optional[OutArray] = None,
+    keepdims: KeepDims = False,
+    initial: NotImplementedType = None,
+    where: NotImplementedType = None,
+):
     axis = _util.allow_only_single_axis(axis)
 
     if dtype == torch.bool:
         dtype = _dtypes_impl.default_dtypes.int_dtype
 
     if axis is None:
-        result = tensor.prod(dtype=dtype)
+        result = a.prod(dtype=dtype)
     else:
-        result = tensor.prod(dtype=dtype, dim=axis)
+        result = a.prod(dtype=dtype, dim=axis)
 
     return result
 
 
-@emulate_keepdims
-@deco_axis_expand
-def mean(tensor, axis=None, dtype=None, *, where=None):
-    dtype = _atleast_float(dtype, tensor.dtype)
+product = prod
+
+
+@_deco_axis_expand
+def mean(
+    a: ArrayLike,
+    axis: AxisLike = None,
+    dtype: Optional[DTypeLike] = None,
+    out: Optional[OutArray] = None,
+    keepdims: KeepDims = False,
+    *,
+    where: NotImplementedType = None,
+):
+    dtype = _atleast_float(dtype, a.dtype)
 
     is_half = dtype == torch.float16
     if is_half:
         # XXX revisit when the pytorch version has pytorch/pytorch#95166
         dtype = torch.float32
 
     if axis is None:
-        result = tensor.mean(dtype=dtype)
+        result = a.mean(dtype=dtype)
     else:
-        result = tensor.mean(dtype=dtype, dim=axis)
+        result = a.mean(dtype=dtype, dim=axis)
 
     if is_half:
         result = result.to(torch.float16)
 
     return result
 
 
-@emulate_keepdims
-@deco_axis_expand
-def std(tensor, axis=None, dtype=None, ddof=0, *, where=None):
-    dtype = _atleast_float(dtype, tensor.dtype)
-    tensor = _util.cast_if_needed(tensor, dtype)
+@_deco_axis_expand
+def std(
+    a: ArrayLike,
+    axis: AxisLike = None,
+    dtype: Optional[DTypeLike] = None,
+    out: Optional[OutArray] = None,
+    ddof=0,
+    keepdims: KeepDims = False,
+    *,
+    where: NotImplementedType = None,
+):
+    dtype = _atleast_float(dtype, a.dtype)
+    tensor = _util.cast_if_needed(a, dtype)
     return tensor.std(dim=axis, correction=ddof)
 
 
-@emulate_keepdims
-@deco_axis_expand
-def var(tensor, axis=None, dtype=None, ddof=0, *, where=None):
-    dtype = _atleast_float(dtype, tensor.dtype)
-    tensor = _util.cast_if_needed(tensor, dtype)
+@_deco_axis_expand
+def var(
+    a: ArrayLike,
+    axis: AxisLike = None,
+    dtype: Optional[DTypeLike] = None,
+    out: Optional[OutArray] = None,
+    ddof=0,
+    keepdims: KeepDims = False,
```

Comment:
In NumPy there are some `keepdims=None` and some `keepdims=False` (you may want to clean this up for NumPy 2.0 @rgommers ?)

I did a search-and-replace of all the `keepdims=None` as it's just unnecessarily annoying having to deal with the `Optional` dtype. I don't expect this to be an issue though, as the code is written so that it works if anyone passes a explicit `None` (who's going to explicitly pass a `None` anyway).

Reply:
And yeah, we should fix the `AxisLike`. Would you mind putting up a PR for this whenever you have time? It's rather low prio tho.

Reply:
AxisLike: yup, will do. Will also need to go through untyped axis arguments, I'm pretty sure I did not do an exemplary job of always normalizing them. At some point.

Reply:
About `Optional`, mostly for my understanding. So the typing-lawyer thing is to only type as `Optional[smth]` things with a None default, not other default values?
[Not that it matters here anyhow]

Reply:
Yeah. Incidentally, that's almost any optional type in NumPy, but well...